src/Main.elm

```diff
@@ -1,37 +1,143 @@
 import Html.Attributes exposing (alt, class, src)
+import Html.Events exposing (onClick)
+import List exposing (append, filter, map, sum)
 
 
 
 ---- MODEL ----
 
 
 type alias Model =
-    { books : List Book }
+    { books : List Book, cart : Cart }
+
+
+type alias CartItem =
+    { id : BookID, quantity : Int }
+
+
+type alias Cart =
+    List CartItem
+
+
+type alias BookID =
+    Int
+
+
+add : Cart -> BookID -> Cart
+add cart bookID =
+    -- let
+    --     _ =
+    --         Debug.log "Logging b" bookID
+    -- in
+    case inCart bookID cart of
+        True ->
+            let
+                incrementQuantity cartItem =
+                    if cartItem.id == bookID then
+                        { cartItem | quantity = cartItem.quantity + 1 }
+
+                    else
+                        cartItem
+            in
+            List.map (\item -> incrementQuantity item) cart
+
+        False ->
+            append cart [ { id = bookID, quantity = 1 } ]
+
+
+remove : Cart -> BookID -> Cart
+remove cart bookID =
+    List.filter (\cartItem -> cartItem.id /= bookID) cart
+
+
+increment : Cart -> BookID -> Cart
+increment cart bookID =
+    let
+        incrementQuantity cartItem =
+            if cartItem.id == bookID then
+                { cartItem | quantity = cartItem.quantity + 1 }
+
+            else
+                cartItem
+    in
+    List.map (\item -> incrementQuantity item) cart
+
+
+decrement : Cart -> BookID -> Cart
+decrement cart bookID =
+    let
+        decrementQuantity cartItem =
+            if cartItem.id == bookID then
+                { cartItem | quantity = cartItem.quantity - 1 }
+
+            else
+                cartItem
+    in
+    List.map (\item -> decrementQuantity item) cart
 
 
 init : ( Model, Cmd Msg )
 init =
-    ( { books = allBooks }
+    ( { books = allBooks, cart = [] }
     , Cmd.none
     )
 
 
+getQuantity : Cart -> BookID -> Int
+getQuantity cart bookID =
+    let
+        result =
+            List.filter (\c -> c.id == bookID) cart
+    in
+    case List.head result of
+        Nothing ->
+            0
+
+        Just c ->
+            c.quantity
+
+
 
 ---- UPDATE ----
 
 
-type Msg
-    = NoOp
+type
+    Msg
+    -- = NoOp
+    = Add BookID
+    | Remove BookID
+    | Increment BookID
+    | Decrement BookID
 
 
 update : Msg -> Model -> ( Model, Cmd Msg )
 update msg model =
-    ( model, Cmd.none )
+    -- ( model, Cmd.none )
+    case msg of
+        Add b ->
+            ( { model | cart = add model.cart b }
+            , Cmd.none
+            )
+
+        Remove b ->
+            ( { model | cart = remove model.cart b }
+            , Cmd.none
+            )
+
+        Increment b ->
+            ( { model | cart = increment model.cart b }
+            , Cmd.none
+            )
+
+        Decrement b ->
+            ( { model | cart = decrement model.cart b }
+            , Cmd.none
+            )
```

Comment:
update function looks super clean! FYI, I think It's okay to do some of the update functionality within this FN, though anything that's not short is a good idea to break out.